[Sky] 미션2 : 체스판 생성 완료

aksmf1442 opened the pull request:
안녕하세요! 자바 체스 미션 2에 대한 PR입니다.

## 변경 사항
* Board, BoardTest 클래스 추가
* color 상수를 enum으로 변경
* BoardTest에서 BeforeEach사용함으로써 중복 제거
* 패키지 분리

리뷰 잘 부탁드립니다.

aksmf1442 committed:
Add Board for chessgame

aksmf1442 committed:
Simplify Pawn, PawnTest

aksmf1442 committed:
Add BoardTest for Board, Update Board

BoardTest(테스트코드)를 먼저 작성하고 이를 기반으로 Board에 필요한 기능들을 구현했다.

aksmf1442 committed:
패키지 나누기

pieces패키지를 생성하고 Pawn, PawnTest클래스를 이 패키지로 이동한다.

aksmf1442 committed:
Board클래스의 pawns를 private으로 변경

Board클래스의 pawns 리스트를 private으로 변경해서 외부로부터의 접근을 막는다.

aksmf1442 committed:
Simplify BoardTest

aksmf1442 committed:
Refactor Pawn color constant to Enum

Pawn클래스에 있는 String타입의 color상수를 enum으로 변경
enum으로 변경함에 따른 사이드 이펙트 수정

SANGCO commented on src/main/java/net/sky/chess/Board.java:
```diff
@@ -0,0 +1,27 @@
+
+public class Board {
+
+  private List<Pawn> pawns;
```
인덴트가 2 spaces로 되어 있는거 같네요?
자바 진영에서는 보통 4 spaces, tab을 쓰고 있습니다.
추가적으로 IntelliJ에 Reformat code 옵션을 찾아서 설정 해봅시다.

SANGCO replied:
이 패드백은 전체 프로젝트에 공통으로 적용입니다~

aksmf1442 replied:
아 그렇군요! 감사합니다. 원래 java-google-style을 쓰고 있었는데 기본 들여쓰기가 2로 되어 있었네요. 수정했습니다!

SANGCO commented on src/main/java/net/sky/pieces/Pawn.java:
```diff
@@ -0,0 +1,21 @@
+
+  private final Color color;
+
+  public enum Color{WHITE, BLACK}
```
이기도 `Color{` 여기 중간에 공백이 필요합니다.
자동 정렬 단축키나 위에서 언급한  IntelliJ에 Reformat code 옵션으로 해결하면 좋을거 같네요.

SANGCO replied:
그리고 아직은 enum에 별다른 기능이 없어서 클래스 안에 선언하신 것도 괜찮은거 같아요.

aksmf1442 replied:
아직 enum에 대한 공부가 많이 안되어 있어서 어떤게 좋은지 안 좋은지를 판별할수가 없네요ㅠㅠ. 더 공부해보겠습니다!

SANGCO commented on src/test/java/net/sky/chess/BoardTest.java:
```diff
@@ -0,0 +1,34 @@
+  public void verifyBoard(final Color color, int size, int idx){
+    Pawn pawn = new Pawn(color);
+    board.add(pawn);
+    assertThat(board.size()).isEqualTo(size);
+    assertThat(board.findPawn(idx)).isEqualTo(pawn);
```
assertAll()에 대해서 학습하고 적용해 보는건 어떨까요~
[JUnit 5 소개](https://javacan.tistory.com/entry/JUnit-5-Intro) (링크에 assertAll() 부분 참고해 주세요.)

aksmf1442 replied:
assertAll()을 사용하면  앞 부분 테스트에서 실패해도 뒤에 부분을 실행해주는군요. 수정했습니다!

SANGCO commented on src/test/java/net/sky/chess/BoardTest.java:
```diff
@@ -0,0 +1,34 @@
+  Board board;
+
+  @BeforeEach
+  public void initBoard(){
```
public 접근 제한자를 제거해도 좋을거 같아요.
JUnit5에서는 테스트 클래스나 메서드에 public을 붙이지 않아도 된답니다.

aksmf1442 replied:
넵. 수정했습니다!

SANGCO commented on src/test/java/net/sky/chess/BoardTest.java:
```diff
@@ -0,0 +1,34 @@
+
+  @Test
+  @DisplayName("체스판 색깔, 사이즈 확인")
+  public void create() throws Exception {
```
체스판 색깔, 사이즈 확인하는 메서드에 이름이 create?
메서드 명에 대해서 조금더 고민해보면 좋을거 같아요~

aksmf1442 replied:
아 저거는 너무 생각없이 메서드명을 짠 것 같네요. 나름대로 수정을 해보긴 했는데도 네이밍이 어려워서 항상 고민되네요ㅠㅠ.

SANGCO replied:
저도 항상 힘들어 하고 있습니다ㅠㅠ

SANGCO commented on src/test/java/net/sky/chess/BoardTest.java:
```diff
@@ -0,0 +1,34 @@
+    assertThat(board.findPawn(idx)).isEqualTo(pawn);
+  }
+
+}
```
![image](https://user-images.githubusercontent.com/26179444/107595669-bab0c180-6c58-11eb-92e5-83782b13ef5b.png)
IntelliJ 설정을 찾아서 적용해 보아요.
[How to make IntelliJ IDEA insert a new line at every end of file?](https://stackoverflow.com/questions/16761227/how-to-make-intellij-idea-insert-a-new-line-at-every-end-of-file)
[파일 끝에 개행을 추가해야 하는 이유](https://blog.coderifleman.com/2015/04/04/text-files-end-with-a-newline/)

aksmf1442 replied:
수정했습니다!

SANGCO commented on src/test/java/net/sky/pieces/PawnTest.java:
```diff
@@ -0,0 +1,31 @@
+
+  @Test
+  @DisplayName("폰 색깔 확인")
+  public void create() {
```
여기도 위에서 남긴 public 접근 제한자, 메서드 명 관련 피드백을 동일하게 적용해 주시면 되겠습니다.

aksmf1442 replied:
이것도 위에 말씀해준 것과 같이 수정했습니다!

SANGCO reviewed (commented):
Sky님 안녕하세요, 리뷰어 상코입니다.
작성하신 코드에 대해서 코멘트를 남겼습니다. 
확인해 보시고 질문 있으시면 댓글 부탁드립니다~

aksmf1442 committed:
Refactor method naming and settings, Use assertAll() for BoardTest

파일 마지막에 개행이 안되어 있던 부분들을 수정
메서드 네이밍 수정
BoardTest 테스트클래스의 verifyBoard라는 메서드에 assertAll()을 사용해서 앞에서 예외가 발생해도 다음 테스트를 실행할 수 있도록 수정
Junit5에서는 굳이 테스트 클래스나 메서드에 public을 붙이지 않아도 된다는 피드백을 받아서 public 제거
